File: browser/lib/peer.ts
import {IMessage, IDescription, ISessionFactory, ISession, IConnectionFactory, IDriver} from "./types";

export default class Peer{
    public message: any;
    constructor(
        private readonly __connect: IConnectionFactory,
        private readonly __session: ISessionFactory,
        private readonly __driver: IDriver){

        this.__driver.on("OFFER", (v,_,s) => this.__onOffer(v as IDescription,s));
        this.__driver.on("ANSWER", (v,_,s) => this.__onAnswer(v as IDescription));
        this.__driver.on("CLOSE", (v,_,s) => this.__onClose(v as IDescription));
    }

    public async connect(id: string,channel: string = "data",option: any = null,...init: Array<(RTCPeerConnection) => any>): Promise<ISession> {
        let conn = this.__connect.create(channel);
        for(let v of init){
            if(!v(conn)){
                conn.close();
                return null;
            }
        }
        
        let session = this.__session.create(id,conn);
        let offer = await session.createOffer(option);
        await session.setLocalDescription(offer);
        return new Promise<ISession>((resolve,reject) => {
            session.onicecandidate = e => {
                if (e.candidate === null) {
                    this.__driver.emitTo(id,"OFFER",{
                        channel,
                        session: session.id(),
                        type: session.localDescription.type,
                        sdp: session.localDescription.sdp,
                    });
                    resolve(session);
                }
            };
        });
    }

    private __create(des: IDescription): ISession{
        let session = this.__session.query(des.session);
        if(!session){
            session = this.__session.attach(des.session,this.__connect.create(des.channel))
        }
        return session;
    }

    private async __onOffer(des: IDescription,source: string){
        const session = this.__create(des);
        await session.setRemoteDescription(new RTCSessionDescription({
            type: "offer",
            sdp: des.sdp,
        }));

        const answer = await session.createAnswer();
        await session.setLocalDescription(answer);

        return new Promise<ISession>((resolve,reject) => {
            session.onicecandidate = e => {
                if (e.candidate === null) {
                    this.__driver.emitTo(source,"ANSWER",{
                        session: des.session,
                        channel: des.channel,
                        type: session.localDescription.type,
                        sdp: session.localDescription.sdp,
                    });
                    resolve();
                }
            };
        });
    }

    private async __onClose(des: IDescription){
        const session = this.__session.query(des.session);
        if(!session){
            session.close();
        }
    }

    private async __onAnswer(des: IDescription){
        const session = this.__session.query(des.session);
        await session.setRemoteDescription(new RTCSessionDescription({
            type: "answer",
            sdp: des.sdp,
        }));

        if((session as any).__resolve){
            (session as any).__resolve(1);
        }
    }
}

File: peer/peer.go
package peer

import (
	"github.com/pion/webrtc/v2"
	"log"
	"sync"
)

type Peer struct {
	driver  Driver
	factory ConnectionFactory
	session SessionFactory
	mutex   sync.Mutex
}

func NewPeer(factory ConnectionFactory, session SessionFactory, driver Driver) *Peer {
	peer := &Peer{
		driver:  driver,
		factory: factory,
		session: session,
	}
	driver.With(peer)
	return peer
}

func (p *Peer) OnClose(id string) (name string, result string, err error) {
	if conn := p.session.Query(id); nil != conn {
		p.session.Remove(id)
		name = "CLOSE"
		result = id
		err = conn.Close()
	}
	return
}

func (p *Peer) OnOffer(offer *Description) (name string, answer *Description, err error) {
	peer, err := p.__create(offer.Session, offer.Channel)
	if nil != err {
		return
	}
	if err = peer.SetRemoteDescription(webrtc.SessionDescription{Type: webrtc.SDPTypeOffer, SDP: offer.SDP}); nil != err {
		log.Printf("SetRemoteDescription error: %s\n", err)
		return
	}
	sdp, err := peer.CreateAnswer(nil)
	if nil == err {
		name = "ANSWER"
		answer = &Description{
			Session:               offer.Session,
			Channel:               offer.Channel,
			SessionDescription: sdp,
		}
	}
	return
}

func (p *Peer) OnAnswer(answer *Description) error {
	if conn := p.session.Query(answer.Session); nil != conn {
		conn.state <- 1
		return conn.SetRemoteDescription(webrtc.SessionDescription{Type: webrtc.SDPTypeAnswer, SDP: answer.SDP})
	}
	return nil
}

func (p *Peer) __create(id string, channel string) (session *Session, err error) {
	if session = p.session.Query(id); nil != session {
		return
	}
	conn, err := p.factory.Create(channel)
	if nil == err {
		session = p.session.Attach(id, conn)
	}
	return
}

func (p *Peer) Connect(id string, channel string, init ...func(conn *webrtc.PeerConnection) error) (*Session, error) {
	conn, err := p.factory.Create(channel)
	if nil != err {
		return nil, err
	}
	for _, v := range init {
		v(conn)
	}
	offer, err := conn.CreateOffer(nil)
	if nil != err {
		conn.Close()
		log.Printf("CreateOffer error: %s\n", err)
		return nil, err
	}

	session := p.session.Create(id, conn)
	p.driver.EmitTo(id, "OFFER", &Description{
		Session:            session.id,
		Channel:            channel,
		SessionDescription: offer,
	})
	return session, nil
}
